cli/open.go

```diff
@@ -0,0 +1,268 @@
+		Handler: func(inv *clibase.Invocation) error {
+			return inv.Command.HelpHandler(inv)
+		},
+		Children: []*clibase.Cmd{
+			r.openVSCode(),
+		},
+	}
+	return cmd
+}
+
+const vscodeDesktopName = "VS Code Desktop"
+
+func (r *RootCmd) openVSCode() *clibase.Cmd {
+	var (
+		generateToken bool
+		testOpenError bool
+	)
+
+	client := new(codersdk.Client)
+	cmd := &clibase.Cmd{
+		Annotations: workspaceCommand,
+		Use:         "vscode <workspace> [<directory in workspace>]",
+		Short:       "Open a workspace in Visual Studio Code",
+		Middleware: clibase.Chain(
+			clibase.RequireRangeArgs(1, 2),
+			r.InitClient(client),
+		),
+		Handler: func(inv *clibase.Invocation) error {
+			ctx, cancel := context.WithCancel(inv.Context())
+			defer cancel()
+
+			// Check if we're inside a workspace, and especially inside _this_
+			// workspace so we can perform path resolution/expansion. Generally,
+			// we know that if we're inside a workspace, `open` can't be used.
+			insideAWorkspace := inv.Environ.Get("CODER") == "true"
+			inWorkspaceName := inv.Environ.Get("CODER_WORKSPACE_NAME") + "." + inv.Environ.Get("CODER_WORKSPACE_AGENT_NAME")
+
+			// We need a started workspace to figure out e.g. expanded directory.
+			// Pehraps the vscode-coder extension could handle this by accepting
+			// default_directory=true, then probing the agent. Then we wouldn't
+			// need to wait for the agent to start.
+			workspaceQuery := inv.Args[0]
+			autostart := true
+			workspace, workspaceAgent, err := getWorkspaceAndAgent(ctx, inv, client, autostart, codersdk.Me, workspaceQuery)
+			if err != nil {
+				return xerrors.Errorf("get workspace and agent: %w", err)
+			}
+
+			workspaceName := workspace.Name + "." + workspaceAgent.Name
+			insideThisWorkspace := insideAWorkspace && inWorkspaceName == workspaceName
+
+			if !insideThisWorkspace {
+				// We could optionally add a flag to skip wait, like with SSH.
+				wait := false
+				for _, script := range workspaceAgent.Scripts {
+					if script.StartBlocksLogin {
+						wait = true
+						break
+					}
+				}
+				err = cliui.Agent(ctx, inv.Stderr, workspaceAgent.ID, cliui.AgentOptions{
+					Fetch:     client.WorkspaceAgent,
+					FetchLogs: client.WorkspaceAgentLogsAfter,
+					Wait:      wait,
+				})
+				if err != nil {
+					if xerrors.Is(err, context.Canceled) {
+						return cliui.Canceled
+					}
+					return xerrors.Errorf("agent: %w", err)
+				}
+
+				// If the ExpandedDirectory was initially missing, it could mean
+				// that the agent hadn't reported it in yet. Retry once.
+				if workspaceAgent.ExpandedDirectory == "" {
+					autostart = false // Don't retry autostart.
+					workspace, workspaceAgent, err = getWorkspaceAndAgent(ctx, inv, client, autostart, codersdk.Me, workspaceName)
+					if err != nil {
+						return xerrors.Errorf("get workspace and agent retry: %w", err)
+					}
+				}
+			}
+
+			directory := workspaceAgent.ExpandedDirectory // Empty unless agent directory is set.
```

Comment:
Would it make sense to set this as an environment variable?

Reply:
Do you mean like overriding the default directory (pseudo-home) that's defined in the workspace? E.g. `CODER_OPEN_VSCODE_DEFAULT_DIRECTORY=...`?

Reply:
Sorry, now that I read this slightly more, the directory is expanded and reported in by the agent. I was thinking that we could maybe avoid the round-trip. Never mind!